### PR TITLE
Adding basic account deletion

### PR DESCRIPTION
**Pull request by nerdsupremacist**
Big changes:

- Challenges, invoices and sponsoring no longer require there to be a nonnullable Sponsor
- All postings, locations, challenges, comments and group messages are now anonymized 
- If a posting has no comments and no challenge fulfilled it will be deleted
- The history of emails is deleted

### CODE REVIEW

File: src/main/java/backend/controller/UserController.kt [hunk 1]
```diff
@@ -145,6 +146,26 @@ class UserController(private val userService: UserService,
         return UserView(user)
     }
 
+    /**
+     * DELETE /user/{id}/
+     * Edits user with given id
+     */
+    @Caching(evict = [(CacheEvict(POSTINGS, allEntries = true)), (CacheEvict(TEAMS, allEntries = true))])
+    @PreAuthorize("isAuthenticated()")
```
Comment (thepiwo) [inline src/main/java/backend/controller/UserController.kt:154]: do we want to allow admins as well to delete users?

File: src/main/java/backend/model/event/Team.kt [hunk 2]
```diff
@@ -193,7 +193,7 @@ class Team : BasicEntity, Blockable {
     }
 
     private fun Sponsoring.toEmailListing(): String {
-        return "<b>Name</b> ${this.sponsor.firstname} ${this.sponsor.lastname} <b>Status</b> ${this.status} <b>Betrag pro km</b> ${this.amountPerKm.display()} <b>Limit</b> ${this.limit.display()} <b>Gereiste KM</b> ${this.team?.getCurrentDistance()} <b>Spendenversprechen</b> ${this.billableAmount().display()}"
+        return "<b>Name</b> ${this.sponsor?.firstname} ${this.sponsor?.lastname} <b>Status</b> ${this.status} <b>Betrag pro km</b> ${this.amountPerKm.display()} <b>Limit</b> ${this.limit.display()} <b>Gereiste KM</b> ${this.team?.getCurrentDistance()} <b>Spendenversprechen</b> ${this.billableAmount().display()}"
```
Comment (thepiwo) [inline src/main/java/backend/model/event/Team.kt:196]: will the string default to `""` or `null`?

Comment (thepiwo) [reply src/main/java/backend/model/event/Team.kt:196]: `""` would be fine for now, `null` would be weird in emails, if it accidentally happens, e.g. with donation reminders after an account was deleted already

File: src/main/java/backend/model/payment/SponsoringInvoice.kt [hunk 3]
```diff
@@ -156,9 +156,14 @@ class SponsoringInvoice : Invoice {
                 return total
             }
             is Sponsor -> listOf(EmailAddress(registeredSponsor!!.email))
-            else -> throw Exception("No sponsor email found for invoice $id")
+            else -> listOf()
         }
     }
+
+    fun removeSponsor() {
```
Comment (thepiwo) [inline src/main/java/backend/model/payment/SponsoringInvoice.kt:163]: sounds like "delete" maybe we should call it "unassignSponsor"

Comment (thepiwo) [reply src/main/java/backend/model/payment/SponsoringInvoice.kt:163]: same for the other "removeSponsor" methods, I did notice as the naming somehow implied for me that the whole sponsor is removed, not only the userdata.

File: src/main/java/backend/model/user/DeletionServiceImpl.kt [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+                .forEach {
+                    it.unlike(user.account)
+                    postingRepository.save(it)
+                }
+
+        postingRepository
+                .findAllCommentedByUser(user.account.id!!)
+                .forEach { deleteComments(it, user) }
+
+        locationRepository
```
Comment (thepiwo) [inline src/main/java/backend/model/user/DeletionServiceImpl.kt:132]: this will change donation sums, I would not do this.

**Review by thepiwo — COMMENTED** [review]
Its okay as a quick-fix, as long as we make sure donation summs are not affected (e.g. by deleting locations)

But I'd prefer a proper anonymization service, to work around deletion.

File: src/main/java/backend/model/user/DeletionServiceImpl.kt (hunk 4 above)
Comment (nerdsupremacist) [reply src/main/java/backend/model/user/DeletionServiceImpl.kt:132]: No it won't change it. The location is still assigned to the team. It's just no longer assigned to the user

File: src/main/java/backend/model/payment/SponsoringInvoice.kt (hunk 3 above)
Comment (nerdsupremacist) [reply src/main/java/backend/model/payment/SponsoringInvoice.kt:163]: Cool I'll rename them

File: src/main/java/backend/model/user/DeletionServiceImpl.kt (hunk 4 above)
Comment (nerdsupremacist) [reply src/main/java/backend/model/user/DeletionServiceImpl.kt:132]: I think I will rename them to `anonymize` to make it clear

File: src/main/java/backend/controller/UserController.kt (hunk 1 above)
Comment (nerdsupremacist) [reply src/main/java/backend/controller/UserController.kt:154]: Right now I'm allowing Admins and the users themselves. We can discuss it, but for now I think we need to have admins on it, just to make sure we shake off all the bugs I wasn't able to see until now

File: src/main/java/backend/model/user/DeletionServiceImpl.kt (hunk 4 above)
Comment (nerdsupremacist) [reply src/main/java/backend/model/user/DeletionServiceImpl.kt:132]: You can see in the picture here: https://github.com/BreakOutEvent/breakout-frontend/pull/367

That it didn't change the donation promise or the score of the team

**Review by thepiwo — APPROVED** [review]
Great Feedback, sounds reasonable

File: src/main/java/backend/model/event/Team.kt (hunk 2 above)
Comment (nerdsupremacist) [reply src/main/java/backend/model/event/Team.kt:196]: I will make an issue to handle those

**Pull request merged**